from fastapi import FastAPI, Depends, HTTPException
from app.configurations.config import SECRET_KEY, ALGORITHM, oauth2_scheme
from app.cogs import models
from sqlalchemy.orm import Session
from jose import JWTError, jwt
from app.cogs.get_db import get_db

def get_current_user(token: str = Depends(oauth2_scheme), db: Session = Depends(get_db)):
    credentials_exception = HTTPException(
        status_code=401,
        detail="Could not validate credentials",
        headers={"WWW-Authenticate": "Bearer"},
    )
    try:
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        email: str = payload.get("sub")
        if email is None:
            raise credentials_exception
    except JWTError:
        raise credentials_exception

    user = db.query(models.User).filter(models.User.email == email).first()
    if user is None:
        raise credentials_exception
    return user